File: codegen/codegen-erlang/src/test/resources/ir/credential_provider_chain.expected.erl
-spec resolve_chain(client_config()) -> {ok, aws_credentials()} | {error, term()}.
resolve_chain(Config) -> resolve_chain(Config, [env, profile, ecs, ec2]).

resolve_chain(_Config, []) ->
    {error, not_found};
resolve_chain(Config, [Provider | Rest]) ->
    case resolve_provider(Provider, Config) of
        {ok, Creds} -> {ok, Creds};
        _ -> resolve_chain(Config, Rest)
    end.

-spec resolve_provider(atom(), client_config()) -> {ok, aws_credentials()} | {error, term()}.
resolve_provider(env, Config) -> resolve_from_env(Config);
resolve_provider(profile, Config) -> resolve_from_profile(Config);
resolve_provider(ecs, Config) -> resolve_from_ecs(Config);
resolve_provider(ec2, Config) -> resolve_from_ec2(Config).

File: codegen/codegen-elixir/src/test/resources/ir/credential_provider_chain.expected.ex
@spec resolve_chain(client_config()) :: {:ok, aws_credentials()} | {:error, term()}
defp resolve_chain(config) do
  resolve_chain(config, [:env, :profile, :ecs, :ec2])
end

defp resolve_chain(_config, []), do: {:error, :not_found}

defp resolve_chain(
  config,
  [provider | rest]
) do
  case resolve_provider(provider, config) do
    {:ok, creds} -> {:ok, creds}

    _ -> resolve_chain(config, rest)
  end
end

defp resolve_provider(:env, config), do: resolve_from_env(config)
defp resolve_provider(:profile, config), do: resolve_from_profile(config)
defp resolve_provider(:ecs, config), do: resolve_from_ecs(config)
defp resolve_provider(:ec2, config), do: resolve_from_ec2(config)